### PR TITLE
feat(save): Friendly commit messages by analyzing head, body, and transform

### PR DESCRIPTION
**Pull request by dustmop**
If a body is small enough, inline it before calling friendly. This is because friendly uses deepdiff, which is not necessarily fast enough for large bodies.

Read transform scripts to see if they changed between commits, letting us have a commit where the transform changes but body stays the same.

Drop derived values from structure before making the friendly message. This is currently less than ideal, since component needs some work to operate on maps.

Diff the head and body separately to get much better estimates for how much of a body changes.

Remove orphaned test private key in the old dataset tests, move it to testPeers.

### CODE REVIEW

**Comment by dustmop** [comment]
Fixes https://github.com/qri-io/qri/issues/1217 and https://github.com/qri-io/qri/issues/1215.

**Comment by dustmop** [comment]
Should also mention, this is the sequel to https://github.com/qri-io/qri/pull/1031.

**Comment by dustmop** [comment]
Performance numbers for this feature. Note that the implementation cuts off body diffing at 20M, so we should expect larger save operations to not be affected.

I used the `.circleci/user_benchmarks.sh` script for this. Timing of `save` action, numbers are for "user","system","total", 4 runs each:

Before (master):

```
Small (10,000):
0.477 0.116 0.578
0.495 0.114 0.782
0.498 0.110 0.594
0.489 0.112 0.580

Medium (50,000):
2.167 0.339 1.952
2.135 0.331 2.010
2.117 0.315 1.886
2.107 0.319 1.911

Large (100,000):
5.050 0.736 4.320
4.240 0.595 3.606
4.290 0.608 3.643
4.294 0.603 3.638
```

After (friendly-commit):

```
Small (10,000):
0.689 0.138 0.796
0.684 0.131 0.783
0.686 0.139 0.799
0.676 0.134 0.791

Medium (50,000):
2.129 0.331 1.918
2.119 0.322 1.892
2.171 0.344 1.950
2.141 0.332 1.920

Large (100,000):
4.584 0.665 3.949
4.489 0.678 3.851
4.380 0.640 3.788
4.290 0.628 3.666
```

I also tried increasing the limit to 200M to see how it affects saving, but only ran 1 time not 4:

```
Small (10,000):
0.682 0.143 0.806

Medium (50,000):
3.254 0.447 3.066

Large (100,000):
6.594 0.956 6.115
```

We can see that including body diffing increasing the time to save by about 50%. I would argue that this is acceptable for small bodies (below 20M) as done in the implementation, but probably too slow after that.

File: base/dsfs/dataset.go [hunk 1]
```diff
@@ -21,10 +22,16 @@ import (
...
 	"github.com/qri-io/qfs/cafs"
+	"github.com/qri-io/qfs/localfs"
 	"github.com/qri-io/qri/base/friendly"
 	"github.com/qri-io/qri/base/toqtype"
 )
 
+var (
+	// BodySizeSmallEnoughToDiff sets how small a body must be to generate a message from it
+	BodySizeSmallEnoughToDiff = 20000000 // 20M or less is small
+)
```
Comment (b5) [inline base/dsfs/dataset.go]: skip parens if only one variable

File: base/dsfs/dataset.go [hunk 2]
```diff
@@ -241,17 +248,30 @@ var Timestamp = func() time.Time {
...
+// BodyAction represents the action that should be taken to understand how the body changed
+type BodyAction string
+
+const (
+	// BodyDefault is the default action: compare them to get how much changed
+	BodyDefault = BodyAction("default")
+	// BodySame means that the bodies are the same, no need to compare
+	BodySame = BodyAction("same")
+	// BodyTooBig means the body is too big to compare, and should be assumed to have changed
+	BodyTooBig = BodyAction("too_big")
```
Comment (b5) [inline base/dsfs/dataset.go]: fan of this. Helps me think about FSI & transform script interactions where a file exists on disk that we don't want to use if unchanged. Don't think we can solve that here, but enumerating what actions we can take is a nice technique

File: base/dsfs/dataset.go [hunk 3]
```diff
@@ -470,16 +509,49 @@ func setChecksumAndLength(ds *dataset.Dataset, data qfs.File, buf *bytes.Buffer,
...
+	if prev.Transform != nil && prev.Transform.ScriptPath != "" {
+		err := prev.Transform.OpenScriptFile(ctx, store)
+		if err == nil {
+			tfFile := prev.Transform.ScriptFile()
+			prev.Transform.ScriptBytes, _ = ioutil.ReadAll(tfFile)
+		}
+	}
+	if ds.Transform != nil && ds.Transform.ScriptPath != "" {
+		// TODO(dustmop): The ipfs filestore won't recognize local filepaths, we need to use
+		// local here. Is there some way to have a cafs store that works with both?
```
Comment (b5) [inline base/dsfs/dataset.go]: yes. Thats covered in the interface that's intended to replace `cafs.Filestore`. 

`qfs.Filesystem` has a `Mux` implementation for exactly this: https://github.com/qri-io/qfs/blob/a6a7b98f0fa0969f6eae50fe14c2c6b5b32f4b4f/mux.go#L13-L17

To get there we'd need to do a _lot_ of refactoring.

File: base/dsfs/dataset.go [hunk 4]
```diff
@@ -509,6 +634,37 @@ func generateCommitDescriptions(prev, ds *dataset.Dataset, force bool) (short, l
 	return shortTitle, longMessage, nil
 }
 
+// copied from base/body.go to avoid circular dependency
```
Comment (b5) [inline base/dsfs/dataset.go]: innnnnnteresting. we should think about reversing the flow of this, moving `ReadAllEntries` into dsfs

File: base/dsfs/dataset_test.go [hunk 5]
```diff
@@ -2,34 +2,25 @@ package dsfs
...
...cBels3nm/fWScvjH4dMQXlavwcwPgKhy2NczDhr4X69oEw6Msd4hQiqJrlWd8juUg6vIsrl1wS/JAOCS65fuyJfV3Pw64rWbTPMwO3FOvxj+rFghZFQgjg/i45uHA2UUkM+h504M5Nzs6Arr/rgV7uPGR5e5OBw3lfiS9ZaA7QZiOq7sMy1L0qD49YO1ojqWu3b7UaMaBQx1Dty7b5IVOSYG+Y3U/dLjhTj4Hg1VtCHWRm3nMOE9cVpMJRhRzKhkq6gnZmni8obz2BBDF02X34oQLcHC/Wn8F3E8RiBjZDI66g+iZeCCUXvYz0vxWAQQKBgQDEJu6flyHPvyBPAC4EOxZAw0zh6SF/r8VgjbKO3n/8d+kZJeVmYnbsLodIEEyXQnr35o2CLqhCvR2kstsRSfRz79nMIt6aPWuwYkXNHQGE8rnCxxyJmxV4S63GczLk7SIn4KmqPlCI08AU0TXJS3zwh7O6e6kBljjPt1mnMgvr3QKBgQD6fAkdI0FRZSXwzygx4uSg47Co6X6ESZ9FDf6ph63lvSK5/eue/ugX6p/olMYq5CHXbLpgM4EJYdRfrH6pwqtBwUJhlh1xI6C48nonnw+oh8YPlFCDLxNG4tq6JVo071qH6CFXCIank3ThZeW5a3ZSe5pBZ8h4bUZ9H8pJL4C7yQKBgFb8SN/+/qCJSoOeOcnohhLMSSD56MAeK7KIxAF1jF5isr1TP+rqiYBtldKQX9bIRY3/8QslM7r88NNj+aAuIrjzSausXvkZedMrkXbHgS/7EAPflrkzTA8fyH10AsLgoj/68mKr5bz34nuY13hgAJUOKNbvFeC9RI5g6eIqYH0FAoGAVqFTXZp12rrK1nAvDKHWRLa6wJCQyxvTU8S1UNi2EgDJ492oAgNTLgJdb8kUiH0CH0lhZCgr9py5IKW94OSM6l72oF2UrS6PRafHC7D9b2IV5Al9lwFO/3MyBrMocapeeyaTcVBnkclz4Qim3OwHrhtFjF1ifhP9DwVRpuIg+dECgYANwlHxLe//tr6BM31PUUrOxP5Y/cj+ydxqM/z6papZFkK6Mvi/vMQQNQkh95GH9zqyC5Z/yLxur4ry1eNYty/9FnuZRAkEmlUSZ/DobhU0Pmj8Hep6JsTuMutref6vCk2n02jc9qYmJuD7iXkdXDSawbEG6f5C4MUkJ38z1t1OjA==`)
```
Comment (b5) [inline base/dsfs/dataset_test.go]: nice pickup

File: base/dsfs/structure.go [hunk 6]
```diff
@@ -8,6 +8,15 @@ import (
 	"github.com/qri-io/qfs/cafs"
 )
 
+// BaseTabularSchema is the base schema for tabular data
```
Comment (b5) [inline base/dsfs/structure.go]: When we start supporting object-of-equal-length-array tables as a format, this'll explode..

We should add a note here that this shouldn't be used. Instead favor methods in `github.com/qri-io/dataset/tabular`. 

The even better thing to do would be to move this into the `tabular` package (not in this PR, some other time) and rename it to `BaseSchemaArray`, like we do with JSON identity schemas

File: base/friendly/friendly.go [hunk 7]
```diff
@@ -63,19 +72,15 @@ func DiffDescriptions(deltas []*deepdiff.Delta, stats *deepdiff.Stats) (string,
...
 					// it for both the long message and short title.
 					msg = fmt.Sprintf("%s:\n\t%s", compName, changes.Rows[0])
 					shortTitle = fmt.Sprintf("%s %s", compName, changes.Rows[0])
-				} else if len(changes.Rows) == 0 && compName == "transform" {
-					// TODO (b5) - this is a hack to make TestSaveTransformWithoutChanges
-					// in the cmd package pass. We're getting to this stage with 0 rows of
-					// changes, which is making msg & title not-empty, which is in turn allowing
-					// a commit b/c it looks like a change. ideally we don't make it here at all,
-					// but we DEFINITELY need a better hueristic for dsfs.CreateDataset's
-					// change detection check. Maybe use diffstat?
```
Comment (b5) [inline base/friendly/friendly.go]: yessssss

File: base/dsfs/dataset.go (hunk 4 above)
Comment (dustmop) [reply base/dsfs/dataset.go]: I agree, will do in a follow-up change.

File: base/dsfs/structure.go (hunk 6 above)
Comment (dustmop) [reply base/dsfs/structure.go]: I used to be able to use dataset's BaseSchemaArray, but some tests were failing that "items" needed to exist at the top-level. I think some lower codepath has started relying on arrays always being 2-dimensional?

**Comment by dustmop** [comment]
Something I forgot to mention, that I added as a TODO and am also copying here. Limiting the body size for diffs causes an undesirable side-effect: if a user has a dataset body larger than this size, then any `save` operation will create a commit, even if no changes have been made, since our change detection will see BodyTooBig and assume something changed. Some solutions to this:
* above 20M start using a less-accurate non-structured checksum instead of deepdiff
* calculate a structured checksum and compare that to the last commit's checksum
* immitate "redo" (https://apenwarr.ca/log/20181113), store file attributes, see if any change
* at a large enough size, like 4G, just assume the file is always changed. don't be expensive
We should make this algorithm agree with how `status` works. See issue: https://github.com/qri-io/qri/issues/1150

**Comment by b5** [comment]
hrm... I didn't catch the proviso for files over `BigFile`. sounds like a job for a rolling hash function.

**Pull request merged**